File: dart/hangman.dart
import 'dart:io';
import 'dart:convert';
import 'dart:math';

String getWord() {
  var rand = Random();
  var file = File('dict_en.txt');
  var words = file.readAsStringSync(encoding: ascii).split('\n');
  var word = '';
  while (word.isEmpty) {
    word = words[rand.nextInt(words.length)].trim();
  }
  return word;
}

Map<Object, Object> getResult(word, guesses) {
  var result = '';
  var incorrect = 0;
  for (var rune in word.runes.toList()) {
    var char=new String.fromCharCode(rune);
    if (guesses.contains(char.toLowerCase())) {
      result += char;
    }
    else {
      result += '_';
      incorrect++;
    }
  };
  return { 'result': result, 'incorrect': incorrect };
}

bool isalpha(c) {
  return 'abcdefghijklmnopqrstuvwxyz'.contains(c.toLowerCase());
}

void main() {
  var maxFalseAttempts = 5;
  var falseAttempts = 0;
  var endOfGame = false;
  var word = getWord();
  var guesses = '';
  print('Guess my secret word with a maximum' +
    'of ${maxFalseAttempts} false attempts.');
  while (!endOfGame) {
    var test = getResult(word, guesses);
    print(test['result']);
    if (test['incorrect'] == 0) {
      endOfGame = true;
      print('You won!');
    }
    else{
      print('Guess a character: ');
      var guess = stdin.readLineSync().toLowerCase();
      if (guess.length != 1) {
        print('Type a single character only! Try again...');
        continue;
      }
      if (!isalpha(guess)) {
        print('Type a alphabetic character! Try again...');
        continue;
      }
      if (guesses.contains(guess)) {
        print('You already guessed this character! Try again...');
        continue;
      }
      guesses += guess;
      if (!word.contains(guess)) {
        falseAttempts++;
        print('Wrong.');
        if (falseAttempts == 1) {
          print('You have one false attempt.');
        }
        else {
          print('You have ${falseAttempts} false attempts.');
        }
        if (falseAttempts > maxFalseAttempts) {
          endOfGame = true;
          print('You lost!');
          print('The secret word has been ' + word.toUpperCase());
        }
      }
    }
  }
}

File: haskell/app/hangman.hs
{-|
  Copyright (c) 2019, Oliver Merkel.
  Please see the AUTHORS file for details.
  All rights reserved.
 
  Use of this code is governed by a
  MIT license that can be found in the LICENSE file.
-}

import Data.Char     (toLower)
import Data.Char     (toUpper)
import System.Random (randomRIO)
import System.IO

fileName :: FilePath
fileName = "dict_en.txt"

maxFalseAttempts = 5 :: Int

data GameState = GameState
    { word    :: String
    , guesses :: String
    , falseAttempts :: Int
    }

data TestResult = TestResult
    { result :: String
    , incorrect :: Int
    }

welcome :: IO ()
welcome = putStrLn $ "Guess my secret word with a maximum of " ++ show maxFalseAttempts ++ " false attempts."

getWord :: IO String
getWord = do
    contents <- readFile fileName
    let words = lines contents
    idx <- randomRIO (0 :: Int, (length words) - 1 )
    let word = map toLower (words !! idx )
    return word

getTestResult :: GameState -> TestResult
getTestResult gameState = do
    let result' = map visibility (word gameState)
    let incorrect' = length $ filter isHidden result'
    TestResult result' incorrect'
    where
        visibility c
            | c `elem` (guesses gameState) = c
            | otherwise = '_'
        isHidden c = c == '_'

getSingleChar :: IO Char
getSingleChar = do
    line <- getLine
    case line of
        [] -> getSingleChar
        (c:_) -> return c

getNextChar :: String -> IO Char
getNextChar guesses' = do
    putStrLn "Guess a character: "
    c <- getSingleChar
    if c `elem` guesses'
    then do
        putStrLn "You already guessed this character! Try again..."
        getNextChar guesses'
    else
        return c

playAgain :: IO Bool
playAgain = do
    putStrLn "Play again? (y/n):"
    answer <- getSingleChar
    return (answer == 'y')

gameInit :: IO GameState
gameInit = do
    welcome
    word <- getWord
    let guesses' = ""
    let falseAttempts' = 0
    return $ GameState word guesses' falseAttempts'

gameLoop :: GameState -> IO ()
gameLoop gameState = do
    let test = getTestResult gameState
    putStrLn (result test)
    if (incorrect test) == 0 then
        putStrLn "You won!"
    else do
        c <- getNextChar (guesses gameState)
        falseAttempts' <- if (c `notElem` (word gameState)) then do
                putStrLn "Wrong."
                if (falseAttempts gameState) == 0 then
                    putStrLn "You have one false attempt."
                else
                    putStrLn $ "You have " ++ show ((falseAttempts gameState) + 1) ++ " false attempts."
                return ((falseAttempts gameState) + 1)
            else
                return (falseAttempts gameState)
        let endOfGame' = (c `notElem` (word gameState)) && falseAttempts' > maxFalseAttempts
        if endOfGame' then do
            putStrLn "You lost!"
            putStrLn $ "The secret word has been " ++ map toUpper (word gameState)
        else
            gameLoop $ gameState { guesses = guesses gameState ++ [c],
                falseAttempts = falseAttempts'}

mainLoop :: IO ()
mainLoop = do
    hSetBuffering stdin NoBuffering
    hSetBuffering stdout NoBuffering
    gameInit >>= gameLoop
    keepPlaying <- playAgain
    if keepPlaying then mainLoop else putStrLn "Bye."

main :: IO ()
main = mainLoop
